### PR TITLE
Julia codegen

### PR DESCRIPTION
**Pull request by p-zubieta**
This is shamelessly adapted from the octave/Matlab code given the syntax similitude between the languages.

It's still a work in progress and any help or recommendation is more than welcomed.

CC @asmeurer @jverzani

### CODE REVIEW

File: doc/src/modules/codegen.rst [hunk 1]
```diff
@@ -59,7 +59,8 @@ This is where the meat of code generation is; the translation of SymPy
 expressions to specific languages. Supported languages are C
 (:py:func:`sympy.printing.ccode.ccode`), Fortran 95
 (:py:func:`sympy.printing.fcode.fcode`), Javascript
-(:py:func:`sympy.printing.jscode.jscode`) , Mathematica
+(:py:func:`sympy.printing.jscode.jscode`), Mathematica
+(:py:func:`sympy.printing.julia.julia_code`), Julia
```
Comment (asmeurer) [inline doc/src/modules/codegen.rst]: These are in the wrong order.

**Comment by asmeurer** [comment]
This needs tests.

**Comment by pabloferz** [comment]
Yeah. I'm working on the tests. I believe that `codegen.py` can use a redisign to reduce redundant code, but I gues that should be left into another PR.

**Comment by asmeurer** [comment]
Any thoughts on how to redesign it to remove all the redundancy?

**Comment by moorepants** [comment]
If Julia is close enough to Octave, can you simply subclass it?

We could also try to push redundant code to an abstract base class.

**Comment by pabloferz** [comment]
I was thinking more on the approach. But as it is commented somewhere in the [CF]CodeGen, it seems that those can be reorganized too, so I was thinking in a more general reorganization.

In the meantime, if writing a parent class for Octave and Julia seems reasonable I can do that here.

**Comment by asmeurer** [comment]
CCing @cbm755 since he worked on the octave codegen and may have thoughts.

**Comment by cbm755** [comment]
1. Some kind of code reuse sounds very good, either by subclassing octave or with a common superclass.  To first approx isn't Julia just `a[1]` instead of `a(1)`?  ;-)

2. I found the codegen stuff rather hard to work with when I was doing Octave: big general refactor probably do-able too---but no need to hold back (1) for that.

**Comment by pabloferz** [comment]
Yes, Julia uses braces for indexing.

I guess I can do (1) above. As I said, (2) should be handled in another PR.

**Comment by cbm755** [comment]
We may also want separate codegen for Matlab and Octave (Matlab is roughly a subset of Octave).

So maybe we want a `OJMCodePrinter(CodePrinter)` with thin subclasses for each of Octave, Julia, and Matlab?

**Comment by cbm755** [comment]
`OJMCodePrinter` `<--` subject to better terminology ;-)

**Comment by cbm755** [comment]
`AlgolLikeCodePrinter`, `QuasiAlgolCodePrinter`?  citation: http://blogs.mathworks.com/cleve/2015/06/15/algol-60-pl0-and-matlab/

**Comment by asmeurer** [comment]
Instead of trying to figure out how to heiracharly order the languages we support, I think it would be better to factor out some of the algorithms in a language agnostic way into the CodePrinter superclass for all the code generating printers. For example, `_print_Mul` here is almost the same as the `CodePrinter._print_Mul`. I didn't look too closely, but it from a glance it seems that one of the main differences is the division and product symbol, which could be factored into a `divsym` attribute (or method) to be overridden on subclasses. 

Basically, subclass hierarchies aren't going to be helpful if you have to copy an entire method just to change one string in that method. If the method is only one line long, that's fine. The ones with complicated logic (like `_print_Mul`) are the main ones to avoid code duplication with, because they are the highest risk of being changed in one place but not in others.

**Comment by asmeurer** [comment]
Also, don't worry about this for this PR. Just adding the new Julia printer as you have done for now is fine.

**Comment by pabloferz** [comment]
Ok. I think this is ready (unless I missed something).

**Comment by jverzani** [comment]
In looking over the tests, I see rationals in SymPy come back as floats in Julia. Is that necessary?

**Comment by pabloferz** [comment]
I guess they could be returned as Julia `Rational`s. I just thought it wouldn't matter too much in a large number of cases. For example, let's say you generate the expression `f(x) = 3*x^2/5 - 5` (where the 3/5 came from a SymPy rational) and use that in Julia to compute f(1//2), then you'll get `-97//20`. I guess there are scenarios where you would really prefer to get rationals.

If you strongly think it is better to always return rationals as rationals I can consider to change it.

**Comment by jverzani** [comment]
On second thought. My aesthetic sense is getting in the way of the use case which is to create fast expressions in julia for sympy expressions. These will almost certainly not involve rational expressions on the julia side. Seems better to keep the exact expressions in sympy.

**Comment by asmeurer** [comment]
I'm going to take your word for it that the Julia code is correct (unless someone else who knows Julia better wants to review this).

File: sympy/printing/tests/test_julia.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+from sympy.matrices import (eye, Matrix, MatrixSymbol, Identity,
+                            HadamardProduct, SparseMatrix)
+from sympy.functions.special.bessel import (jn, yn, besselj, bessely, besseli,
+                                            besselk, hankel1, hankel2, airyai,
+                                            airybi, airyaiprime, airybiprime)
+from sympy.utilities.pytest import XFAIL
+from sympy.core.compatibility import range
+
+from sympy import julia_code
+from sympy import julia_code as mcode
```
Comment (asmeurer) [inline sympy/printing/tests/test_julia.py]: This is confusing. I'm guessing this was done because you copied this from the octave printer. I'd rather just test against "julia_code".

**Comment by pabloferz** [comment]
@jverzani might have something to say about it if you want a second pair of eyes.

File: sympy/printing/tests/test_julia.py (hunk 2 above)
Comment (pabloferz) [inline sympy/printing/tests/test_julia.py]: Fair enough. I'll change it.

**Comment by asmeurer** [comment]
If no one else has any comments I'm +1 to merge this.

**Comment by asmeurer** [comment]
Tests fail. Looks like you missed one. Be sure to run `./bin/test julia` before committing.

**Comment by pabloferz** [comment]
Yeah, sorry. It is fixed now, and I have tested locally before pushing.
